### ISSUE TITLE
Airflow worker pods starts very slow because of K8s init container

### ISSUE THREAD

**Issue by jagusztinl** [bitnami/charts#5651]
**Which chart**:
The name (and version) of the affected chart
bitnami/airflow 2.0.1
**Describe the bug**

Airflow worker pods starts very slow because of K8s init container

checked the container log:
�[37mnami �[39m �[32mINFO �[39m Initializing airflow-worker
�[36mairflow�[39m �[32mINFO �[39m ==> No injected configuration files found. Creating default config file...

!!!!!!30 seconds pause here!!!!!!!

�[36mairflow�[39m �[32mINFO �[39m ==> Configuring Airflow database...
�[36mairflow�[39m �[32mINFO �[39m ==> Setting Airflow Fernet key...
�[36mairflow�[39m �[32mINFO �[39m ==> Deploying Airflow from scratch...
�[36mpostgre�[39m �[32mINFO �[39m Trying to connect to PostgreSQL server
�[36mpostgre�[39m �[32mINFO �[39m Found PostgreSQL server listening at airflow-postgresql:5432
�[36mpostgre�[39m �[32mINFO �[39m PostgreSQL server listening and working at airflow-postgresql:5432
�[36mairflow�[39m �[32mINFO �[39m ==> Waiting for Airflow Webserver host...
�[36mairflow�[39m �[32mINFO �[39m
�[36mairflow�[39m �[32mINFO �[39m ########################################################################
�[36mairflow�[39m �[32mINFO �[39m Installation parameters for airflow-worker:
�[36mairflow�[39m �[32mINFO �[39m Webserver URL: airflow:8080
�[36mairflow�[39m �[32mINFO �[39m This installation requires no credentials.
�[36mairflow�[39m �[32mINFO �[39m ########################################################################
�[36mairflow�[39m �[32mINFO �[39m
�[37mnami �[39m �[32mINFO �[39m airflow-worker successfully initialized

How can I make it faster?

**Expected behavior**
Start workflow pods much faster

**Version of Helm and Kubernetes**:

- Output of `helm version`:

```
(paste your output here)
```

- Output of `kubectl version`:

```
(paste your output here)
```

**Additional context**

**Comment by miguelaeh** [comment]
Hi @jagusztinl ,
As you can see in the log:
>�[36mairflow�[39m �[32mINFO �[39m ==> No injected configuration files found. Creating default config file...

During that time it is creating the default configuration. That should only happen the first time when the volumes are empty, for future restarts it should not wait for that long since the configuration will be there.

**Comment by cris-b** [comment]
> Hi @jagusztinl ,
> As you can see in the log:
> 
> > �[36mairflow�[39m �[32mINFO �[39m ==> No injected configuration files 
>  That should only happen the first time when the volumes are empty, for future restarts it should not wait for that long since the configuration will be there.

But if using kubernetes executor the pods are required created for every task, so this init container runs every time a task runs, meaning this may be the longest part of the job..

**Comment by miguelaeh** [comment]
ah, I see, the volume seems to be always an empty dir when using Kubernetes Executor: https://github.com/miguelaeh/bitnami-charts/blob/b51a8a718bf62f01f0ef4e8f5b0630fd3e2a7b01/bitnami/airflow/templates/config/configmap.yaml#L188

It seems that the init container just initializes and configure airflow to copy the configuration file to the volume to be used by the real container, and that `nami initialize` step is the one taking such time.
To reduce that time, we should find a way to generate that Kubernetes executor configuration without the nami initialize.

I was thinking about how to persist that configuration, but I am not sure right now about the implications of persisting that configuration, it does not seem to be a good idea since the executor could require different configurations depending on the executor.

**Comment by cris-b** [comment]
it seems to me that what is taking the time is that nami initialize is actually running db,init and db,reset - here I have enabled NAMI_DEBUG:

airflow INFO  ==> No injected configuration files found. Creating default config file...
airflow TRACE [runProgram] Executing: /opt/bitnami/airflow/venv/bin/airflow db,init
airflow TRACE [runProgram] Executing: /opt/bitnami/airflow/venv/bin/airflow db,reset

i've no idea why it would need to do this just to create the config (even just running airflow version creates the config), and I cant find what is controlling it.  I see no airflow module for nami anywhere in the image.

**Comment by cris-b** [comment]
ah found it, its in:

/.nami/components/com.bitnami.airflow-worker/helpers.js

$app.helpers.generateConfFile = function() {
  $os.runProgram(
    'venv/bin/airflow',
    this.getAirflowCommand('initdb', 'db init'),
    {env: {AIRFLOW_HOME: $app.installdir}}
  );
  $os.runProgram(
    'venv/bin/airflow',
    this.getAirflowCommand('resetdb', 'db reset'),
    {input: 'y', env: {AIRFLOW_HOME: $app.installdir}}
  );
  $file.delete($file.join($app.installation.root, 'airflow/airflow.db'));
};

so it seems like this function runs the init/reset purely to get the default config file, it even deletes the created db at the end!  Looks like this helper could just run 'airflow version'.

**Comment by cris-b** [comment]
i've now worked around this by extending the worker container and patching it:

dockerfile:

FROM bitnami/airflow-worker:2.0.1-debian-10-r24

COPY helper.patch /

USER root

RUN apt-get update -y && \
    apt-get install -y patch && \
    cd /.nami/components/com.bitnami.airflow-worker && \
    patch -p0 < /helper.patch && \
    apt-get remove -y patch 

USER 1001

patch:

--- helpers.js.orig     2021-03-08 12:06:21.408149521 +0000
+++ helpers.js  2021-03-08 12:06:37.018681863 +0000
@@ -21,15 +21,9 @@
 $app.helpers.generateConfFile = function() {
   $os.runProgram(
     'venv/bin/airflow',
-    this.getAirflowCommand('initdb', 'db init'),
+    this.getAirflowCommand('version', 'version'),
     {env: {AIRFLOW_HOME: $app.installdir}}
   );
-  $os.runProgram(
-    'venv/bin/airflow',
-    this.getAirflowCommand('resetdb', 'db reset'),
-    {input: 'y', env: {AIRFLOW_HOME: $app.installdir}}
-  );
-  $file.delete($file.join($app.installation.root, 'airflow/airflow.db'));
 };
 
i'm not sure helper.js is open source, I cant find a repo for it so publishing the patch here.  Hope this helps!

**Comment by miguelaeh** [comment]
Hi @cris-b ,
Those files are not public on GitHub, that's why we are migrating all the containers to bash instead of nami, but the migration is still in progress.

About the patch, if should be enough, but then you need to use different images for the workers and the airflow deployment, since the initialize is needed to configure the application properly. I guess the reason of execute the full nami initialize is that the `nami` tool does not export any other function to run for example the `version` like you mention.

**Comment by cris-b** [comment]
the chart uses a separate worker image by default anyway.  using nami initialize is fine, its just that what gets called due to nami initialize is not required.

i believe the above patch should be made to all 3 of the bitnami packages (airflow, airflow-scheduler, airflow-worker), because it is the same in each.  The function it is part of (generateConfFile) is run before the db is configured, so it will only ever be run using the local sqlite db.   

In the airflow (webserver) image the bitnami airflow package runs a separate initdb as part of its startup, after the correct db parameters have been initialised, and after generateConfFile has been run. 

Basically the issue in generateConfFile slows the startup of all three images and adds nothing as far as I can tell.

**Comment by miguelaeh** [comment]
Hi @cris-b ,
I am currently working on moving the airflow logic to bash instead of nami, I will have that into account and will test it properly.
